File: network/src/main/java/com/securechat/network/JanusClient.kt
package com.securechat.network

import android.util.Log
import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import okhttp3.*
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * Janus Gateway WebSocket istemcisi.
 * VideoRoom plugin uzerinden SFU modunda grup arama yonetimi saglar.
 *
 * Akis:
 * 1. connect() → Janus WS bagla
 * 2. createSession() → Janus session al
 * 3. attachVideoRoom() → VideoRoom plugin'ine baglan (handle al)
 * 4. joinAsPublisher() → Odaya publisher olarak katil, mevcut publisher listesi al
 * 5. publishSdp() → Yerel SDP offer'i gonder, Janus'tan answer al
 * 6. subscribeToFeed() → Uzak publisher'a abone ol, Janus'tan offer al
 * 7. answerSubscription() → Subscriber SDP answer'i gonder
 */
class JanusClient(
    /**
     * Shared OkHttpClient — NetworkModule'den inject edilir (certificate pinning + interceptors).
     * JanusClient kendi OkHttp instance'ini olusturmaz, aksi takdirde MITM korumasi atlanir.
     */
    private val sharedClient: OkHttpClient
) {

    companion object {
        private const val TAG = "JanusClient"
        private const val JANUS_TIMEOUT_MS = 10_000L

        /**
         * SDP icindeki private/local IP candidate'lari ve mDNS host'lari kaldirir (M6 fix).
         * RFC 1918 araliklar:  10.0.0.0/8, 172.16-31.x.x, 192.168.0.0/16,
         * Link-local: 169.254.0.0/16, IPv6 fe80::/10
         * mDNS host candidate'lari (.local hostname) — gercek IP'yi gizler ama bilgi sizdirir.
         *
         * Sonuc: sadece public/STUN/TURN relay candidate'lari Janus'a iletilir → LAN topology leak yok.
         */
        internal fun stripPrivateCandidates(sdp: String): String {
            val privateCandidateRegex = Regex(
                """^a=candidate:.*\s(?:10\.\d+\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+|192\.168\.\d+\.\d+|169\.254\.\d+\.\d+|fe80:[0-9a-fA-F:]+|[A-Fa-f0-9.-]+\.local)\s.*$""",
                RegexOption.IGNORE_CASE
            )
            return sdp.lineSequence()
                .filterNot { privateCandidateRegex.matches(it) }
                .joinToString("\r\n")
        }
    }

    private var webSocket: WebSocket? = null
    // Janus icin daha uzun read/write timeout — uzun suren VideoRoom operasyonlari icin.
    // newBuilder() shared client'in pinner+interceptor'larini KORUR.
    private val client = sharedClient.newBuilder()
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private var sessionId: Long = 0
    private var publisherHandleId: Long = 0
    private var roomId: Long = 0

    // Subscriber handle'lari: feedId -> handleId
    private val subscriberHandles = ConcurrentHashMap<Long, Long>()

    // Transaction ID -> CompletableDeferred response
    private val pendingRequests = ConcurrentHashMap<String, CompletableDeferred<JsonObject>>()

    // Event callback'leri
    var onPublisherJoined: ((feedId: Long, displayName: String?) -> Unit)? = null
    var onPublisherLeft: ((feedId: Long) -> Unit)? = null
    var onRemoteSdpOffer: ((feedId: Long, sdp: String) -> Unit)? = null
    var onRemoteSdpAnswer: ((sdp: String) -> Unit)? = null

    @Volatile
    var isConnected = false
        private set

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /**
     * Janus WebSocket sunucusuna baglanir.
     *
     * GUVENLIK: api_secret tabanli auth KALDIRILDI (C2 fix).
     * Janus authentication artik Nginx reverse proxy katmaninda yapilir
     * (Authorization: Bearer JWT). Client mesajlarinda apisecret alani gonderilmez.
     */
    /** Eski apiSecret auth no-op — backward compat icin korundu, hicbir sey yapmaz. */
    @Deprecated("Janus auth Nginx reverse proxy katmaninda yapilir; apiSecret kullanmayin.")
    fun setApiSecret(@Suppress("UNUSED_PARAMETER") secret: String?) {
        // No-op: apisecret artik kullanilmiyor.
    }

    /** JsonObjectBuilder auth — Nginx katmani Authorization header'i set eder, body'de degisiklik yok. */
    private fun JsonObjectBuilder.withAuth() {
        // apisecret artik gonderilmez.
    }

    suspend fun connect(janusWsUrl: String): Boolean {
        val connectDeferred = CompletableDeferred<Boolean>()

        val request = Request.Builder()
            .url(janusWsUrl)
            .addHeader("Sec-WebSocket-Protocol", "janus-protocol")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "Janus WS baglandi: $janusWsUrl")
                isConnected = true
                connectDeferred.complete(true)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleJanusMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "Janus WS hatasi: ${t.message}")
                isConnected = false
                connectDeferred.complete(false)
                // Bekleyen tum request'leri iptal et
                pendingRequests.forEach { (_, deferred) ->
                    deferred.completeExceptionally(t)
                }
                pendingRequests.clear()
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "Janus WS kapandi: $reason")
                isConnected = false
            }
        })

        return withTimeout(JANUS_TIMEOUT_MS) { connectDeferred.await() }
    }

    /**
     * Janus session olusturur.
     */
    suspend fun createSession(): Long {
        val txId = newTransaction()
        val msg = buildJsonObject {
            put("janus", "create")
            put("transaction", txId)
            withAuth()
        }
        val response = sendAndWait(txId, msg)
        sessionId = response["data"]?.jsonObject?.get("id")?.jsonPrimitive?.long
            ?: throw RuntimeException("Janus session ID alinamadi")
        Log.d(TAG, "Janus session olusturuldu: $sessionId")

        // Keepalive basalt
        startKeepalive()

        return sessionId
    }

    /**
     * VideoRoom plugin'ine baglanir (attach).
     */
    suspend fun attachVideoRoom(): Long {
        val txId = newTransaction()
        val msg = buildJsonObject {
            put("janus", "attach")
            put("session_id", sessionId)
            put("plugin", "janus.plugin.videoroom")
            put("transaction", txId)
            withAuth()
        }
        val response = sendAndWait(txId, msg)
        publisherHandleId = response["data"]?.jsonObject?.get("id")?.jsonPrimitive?.long
            ?: throw RuntimeException("Janus handle ID alinamadi")
        Log.d(TAG, "VideoRoom plugin baglandi: handle=$publisherHandleId")
        return publisherHandleId
    }

    /**
     * Odaya publisher olarak katilir.
     * @return Mevcut publisher listesi (feedId, displayName)
     */
    suspend fun joinAsPublisher(roomId: Long, displayName: String): List<Pair<Long, String?>> {
        this.roomId = roomId
        val txId = newTransaction()
        val body = buildJsonObject {
            put("request", "join")
            put("room", roomId)
            put("ptype", "publisher")
            put("display", displayName)
        }
        val msg = buildJsonObject {
            put("janus", "message")
            put("session_id", sessionId)
            put("handle_id", publisherHandleId)
            put("transaction", txId)
            put("body", body)
            withAuth()
        }
        val response = sendAndWait(txId, msg)

        // Mevcut publisher listesini parse et
        val pluginData = response["plugindata"]?.jsonObject?.get("data")?.jsonObject
        val publishers = pluginData?.get("publishers")?.jsonArray ?: return emptyList()
        return publishers.map { pub ->
            val obj = pub.jsonObject
            val feedId = obj["id"]?.jsonPrimitive?.long ?: 0
            val display = obj["display"]?.jsonPrimitive?.contentOrNull
            feedId to display
        }.also {
            Log.d(TAG, "Odaya katildi, ${it.size} mevcut publisher")
        }
    }

    /**
     * Yerel SDP offer'i Janus'a gonderir ve SDP answer alir.
     * Publisher olarak medya yayinlamak icin kullanilir.
     */
    suspend fun publishSdp(sdpOffer: String): String {
        val txId = newTransaction()
        // GUVENLIK (M6 fix): SDP'den private/local IP candidate'lari kaldir.
        val sanitizedOffer = stripPrivateCandidates(sdpOffer)
        val body = buildJsonObject {
            put("request", "configure")
            put("audio", true)
            put("video", true)
        }
        val jsep = buildJsonObject {
            put("type", "offer")
            put("sdp", sanitizedOffer)
        }
        val msg = buildJsonObject {
            put("janus", "message")
            put("session_id", sessionId)
            put("handle_id", publisherHandleId)
            put("transaction", txId)
            put("body", body)
            put("jsep", jsep)
            withAuth()
        }
        val response = sendAndWait(txId, msg)
        val answerSdp = response["jsep"]?.jsonObject?.get("sdp")?.jsonPrimitive?.content
            ?: throw RuntimeException("Janus SDP answer alinamadi")
        Log.d(TAG, "Publish SDP answer alindi (${answerSdp.length} byte)")
        return answerSdp
    }

    /**
     * Uzak bir publisher'a abone olur (subscriber).
     * Janus SDP offer gonderir — client answer ile yanit verir.
     *
     * @return SDP offer from Janus
     */
    suspend fun subscribeToFeed(roomId: Long, feedId: Long): String {
        // Yeni subscriber handle attach et
        val handleId = attachSubscriberHandle()
        subscriberHandles[feedId] = handleId

        val txId = newTransaction()
        val body = buildJsonObject {
            put("request", "join")
            put("room", roomId)
            put("ptype", "subscriber")
            put("feed", feedId)
        }
        val msg = buildJsonObject {
            put("janus", "message")
            put("session_id", sessionId)
            put("handle_id", handleId)
            put("transaction", txId)
            put("body", body)
            withAuth()
        }
        val response = sendAndWait(txId, msg)
        val offerSdp = response["jsep"]?.jsonObject?.get("sdp")?.jsonPrimitive?.content
            ?: throw RuntimeException("Janus subscriber SDP offer alinamadi")
        Log.d(TAG, "Subscriber SDP offer alindi: feedId=$feedId (${offerSdp.length} byte)")
        return offerSdp
    }

    /**
     * Subscriber SDP answer'i Janus'a gonderir.
     */
    suspend fun answerSubscription(feedId: Long, sdpAnswer: String) {
        val handleId = subscriberHandles[feedId]
            ?: throw RuntimeException("Subscriber handle bulunamadi: feedId=$feedId")

        val txId = newTransaction()
        // GUVENLIK (M6 fix): SDP'den private/local IP candidate'lari kaldir.
        val sanitizedAnswer = stripPrivateCandidates(sdpAnswer)
        val body = buildJsonObject {
            put("request", "start")
            put("room", roomId)
        }
        val jsep = buildJsonObject {
            put("type", "answer")
            put("sdp", sanitizedAnswer)
        }
        val msg = buildJsonObject {
            put("janus", "message")
            put("session_id", sessionId)
            put("handle_id", handleId)
            put("transaction", txId)
            put("body", body)
            put("jsep", jsep)
            withAuth()
        }
        sendAndWait(txId, msg)
        Log.d(TAG, "Subscriber answer gonderildi: feedId=$feedId")
    }

    /**
     * Janus'a ICE candidate trickle gonderir.
     * handleType: "publisher" -> publisherHandleId; "subscriber:<feedId>" -> subscriberHandles[feedId]
     */
    fun trickleIce(handleId: Long, sdpMid: String?, sdpMLineIndex: Int, candidate: String) {
        // GUVENLIK (M6 fix): Private/local IP candidate'i Janus'a iletme.
        // Tek satirlik candidate string'i regex ile dogrula.
        val privateCandidateRegex = Regex(
            """.*\s(?:10\.\d+\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+|192\.168\.\d+\.\d+|169\.254\.\d+\.\d+|fe80:[0-9a-fA-F:]+|[A-Fa-f0-9.-]+\.local)\s.*""",
            RegexOption.IGNORE_CASE
        )
        if (privateCandidateRegex.matches(candidate)) {
            Log.d(TAG, "Private candidate filtrelendi (M6 fix)")
            return
        }
        val txId = newTransaction()
        val candObj = buildJsonObject {
            put("candidate", candidate)
            sdpMid?.let { put("sdpMid", it) }
            put("sdpMLineIndex", sdpMLineIndex)
        }
        val msg = buildJsonObject {
            put("janus", "trickle")
            put("session_id", sessionId)
            put("handle_id", handleId)
            put("transaction", txId)
            put("candidate", candObj)
            withAuth()
        }
        // Trickle async — yanit beklenmez
        webSocket?.send(msg.toString())
    }

    fun trickleIceCompleted(handleId: Long) {
        val txId = newTransaction()
        val msg = buildJsonObject {
            put("janus", "trickle")
            put("session_id", sessionId)
            put("handle_id", handleId)
            put("transaction", txId)
            put("candidate", buildJsonObject { put("completed", true) })
            withAuth()
        }
        webSocket?.send(msg.toString())
    }

    /** Publisher handle ID'sini doner — trickle ICE icin. */
    fun getPublisherHandleId(): Long = publisherHandleId

    /** Subscriber handle ID'sini doner — trickle ICE icin. */
    fun getSubscriberHandleId(feedId: Long): Long? = subscriberHandles[feedId]

    /** VideoRoom'dan ayrilma — best-effort. */
    suspend fun leaveRoom() {
        if (publisherHandleId == 0L || sessionId == 0L) return
        val txId = newTransaction()
        val body = buildJsonObject {
            put("request", "leave")
        }
        val msg = buildJsonObject {
            put("janus", "message")
            put("session_id", sessionId)
            put("handle_id", publisherHandleId)
            put("transaction", txId)
            put("body", body)
            withAuth()
        }
        try { sendAndWait(txId, msg) } catch (_: Exception) { /* best-effort */ }
    }

    /**
     * Subscriber icin yeni handle attach eder.
     */
    private suspend fun attachSubscriberHandle(): Long {
        val txId = newTransaction()
        val msg = buildJsonObject {
            put("janus", "attach")
            put("session_id", sessionId)
            put("plugin", "janus.plugin.videoroom")
            put("transaction", txId)
            withAuth()
        }
        val response = sendAndWait(txId, msg)
        val handleId = response["data"]?.jsonObject?.get("id")?.jsonPrimitive?.long
            ?: throw RuntimeException("Subscriber handle ID alinamadi")
        Log.d(TAG, "Subscriber handle olusturuldu: $handleId")
        return handleId
    }

    /**
     * Janus session'i canli tutmak icin periyodik keepalive gonderir.
     */
    private fun startKeepalive() {
        scope.launch {
            while (isActive && isConnected) {
                delay(25_000) // Janus varsayilan session timeout 60sn
                if (!isConnected) break
                val msg = buildJsonObject {
                    put("janus", "keepalive")
                    put("session_id", sessionId)
                    put("transaction", newTransaction())
                    withAuth()
                }
                webSocket?.send(msg.toString())
            }
        }
    }

    /**
     * Gelen Janus mesajini isler.
     * Transaction-based response'lar ilgili CompletableDeferred'a yonlendirilir.
     * Asenkron event'ler (publisher join/leave) callback'lere iletilir.
     */
    private fun handleJanusMessage(text: String) {
        try {
            val json = Json.parseToJsonElement(text).jsonObject
            val janus = json["janus"]?.jsonPrimitive?.contentOrNull

            // Transaction-based response
            val txId = json["transaction"]?.jsonPrimitive?.contentOrNull
            if (txId != null && pendingRequests.containsKey(txId)) {
                val deferred = pendingRequests.remove(txId)
                if (janus == "error") {
                    val errorMsg = json["error"]?.jsonObject?.get("reason")?.jsonPrimitive?.contentOrNull ?: "Unknown error"
                    deferred?.completeExceptionally(RuntimeException("Janus error: $errorMsg"))
                } else {
                    deferred?.complete(json)
                }
                return
            }

            // Asenkron event'ler
            when (janus) {
                "event" -> handleEvent(json)
                "webrtcup" -> Log.d(TAG, "WebRTC baglantisi kuruldu")
                "hangup" -> Log.d(TAG, "Janus hangup alindi")
                "detached" -> Log.d(TAG, "Handle detach edildi")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Janus mesaj parse hatasi: ${e.message}")
        }
    }

    /**
     * Janus plugin event'lerini isler.
     * - publishers: Yeni publisher katildi
     * - unpublished/leaving: Publisher ayrildi
     * - jsep (subscriber icin): Janus'tan SDP offer
     */
    private fun handleEvent(json: JsonObject) {
        val pluginData = json["plugindata"]?.jsonObject?.get("data")?.jsonObject ?: return

        // Yeni publisher'lar
        val publishers = pluginData["publishers"]?.jsonArray
        if (publishers != null) {
            for (pub in publishers) {
                val obj = pub.jsonObject
                val feedId = obj["id"]?.jsonPrimitive?.long ?: continue
                val display = obj["display"]?.jsonPrimitive?.contentOrNull
                Log.d(TAG, "Yeni publisher: feedId=$feedId, display=$display")
                onPublisherJoined?.invoke(feedId, display)
            }
        }

        // Publisher ayrildi
        val unpublished = pluginData["unpublished"]?.jsonPrimitive?.longOrNull
        if (unpublished != null) {
            Log.d(TAG, "Publisher ayrildi: feedId=$unpublished")
            subscriberHandles.remove(unpublished)
            onPublisherLeft?.invoke(unpublished)
        }

        val leaving = pluginData["leaving"]?.jsonPrimitive?.longOrNull
        if (leaving != null) {
            Log.d(TAG, "Katilimci ayrildi: feedId=$leaving")
            subscriberHandles.remove(leaving)
            onPublisherLeft?.invoke(leaving)
        }

        // Subscriber icin Janus'tan SDP offer (async event olarak gelebilir)
        val jsep = json["jsep"]?.jsonObject
        if (jsep != null) {
            val type = jsep["type"]?.jsonPrimitive?.contentOrNull
            val sdp = jsep["sdp"]?.jsonPrimitive?.contentOrNull
            if (type == "offer" && sdp != null) {
                // Hangi feed'e ait oldugunu bul
                val senderHandleId = json["sender"]?.jsonPrimitive?.longOrNull
                if (senderHandleId != null) {
                    val feedId = subscriberHandles.entries.find { it.value == senderHandleId }?.key
                    if (feedId != null) {
                        onRemoteSdpOffer?.invoke(feedId, sdp)
                    }
                }
            } else if (type == "answer" && sdp != null) {
                onRemoteSdpAnswer?.invoke(sdp)
            }
        }
    }

    /**
     * Mesaj gonderir ve yanit bekler.
     */
    private suspend fun sendAndWait(txId: String, msg: JsonObject): JsonObject {
        val deferred = CompletableDeferred<JsonObject>()
        pendingRequests[txId] = deferred
        val sent = webSocket?.send(msg.toString()) ?: false
        if (!sent) {
            pendingRequests.remove(txId)
            throw RuntimeException("Janus mesaj gonderilemedi")
        }
        return withTimeout(JANUS_TIMEOUT_MS) { deferred.await() }
    }

    private fun newTransaction(): String = UUID.randomUUID().toString().take(12)

    /**
     * Tum kaynaklari temizler ve baglanitiyi kapatir.
     */
    fun disconnect() {
        scope.cancel()
        isConnected = false
        pendingRequests.forEach { (_, deferred) ->
            deferred.cancel()
        }
        pendingRequests.clear()
        subscriberHandles.clear()
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        sessionId = 0
        publisherHandleId = 0
        onPublisherJoined = null
        onPublisherLeft = null
        onRemoteSdpOffer = null
        onRemoteSdpAnswer = null
        Log.d(TAG, "Janus client disconnect edildi")
    }
}

File: flutter_securechat/lib/src/media/janus_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:web_socket_channel/io.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

sealed class JanusEvent {
  const JanusEvent();
}

class JanusPublisherJoined extends JanusEvent {
  const JanusPublisherJoined(this.feedId, this.displayName);
  final int feedId;
  final String? displayName;
}

class JanusPublisherLeft extends JanusEvent {
  const JanusPublisherLeft(this.feedId);
  final int feedId;
}

class JanusRemoteOffer extends JanusEvent {
  const JanusRemoteOffer(this.feedId, this.sdp);
  final int feedId;
  final String sdp;
}

class JanusRemoteAnswer extends JanusEvent {
  const JanusRemoteAnswer(this.sdp);
  final String sdp;
}

class JanusClient {
  JanusClient({
    this.requestTimeout = const Duration(seconds: 10),
    HttpClient? httpClient,
  }) : _httpClient = httpClient;

  final Duration requestTimeout;
  final HttpClient? _httpClient;
  final _events = StreamController<JanusEvent>.broadcast();
  final _pending = <String, Completer<Map<String, Object?>>>{};
  final _subscriberHandles = <int, int>{};
  final Random _random = Random.secure();
  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _subscription;
  Timer? _keepalive;
  int _sessionId = 0;
  int _publisherHandleId = 0;
  int _roomId = 0;
  bool _connected = false;
  bool _disposed = false;
  Future<void>? _disposeTask;

  Stream<JanusEvent> get events => _events.stream;
  bool get isConnected => _connected;
  int get publisherHandleId => _publisherHandleId;
  int? subscriberHandleId(int feedId) => _subscriberHandles[feedId];

  Future<bool> connect({
    required String url,
    required String accessToken,
  }) async {
    if (_disposed) throw StateError('Janus client is disposed');
    await disconnect();
    try {
      final channel = IOWebSocketChannel.connect(
        Uri.parse(url),
        protocols: const ['janus-protocol'],
        headers: {'Authorization': 'Bearer $accessToken'},
        connectTimeout: requestTimeout,
        pingInterval: const Duration(seconds: 20),
        customClient: _httpClient,
      );
      _channel = channel;
      _subscription = channel.stream.listen(
        _handleMessage,
        onError: _handleSocketError,
        onDone: _handleSocketDone,
      );
      await channel.ready.timeout(requestTimeout);
      _connected = true;
      return true;
    } catch (_) {
      await disconnect();
      return false;
    }
  }

  Future<int> createSession() async {
    final response = await _request({'janus': 'create'});
    _sessionId = _intAt(response, ['data', 'id']);
    if (_sessionId == 0) throw StateError('Janus session id is missing');
    _keepalive = Timer.periodic(const Duration(seconds: 25), (_) {
      if (!_connected || _sessionId == 0) return;
      _send({
        'janus': 'keepalive',
        'session_id': _sessionId,
        'transaction': _transaction(),
      });
    });
    return _sessionId;
  }

  Future<int> attachVideoRoom() async {
    _requireSession();
    final response = await _request({
      'janus': 'attach',
      'session_id': _sessionId,
      'plugin': 'janus.plugin.videoroom',
    });
    _publisherHandleId = _intAt(response, ['data', 'id']);
    if (_publisherHandleId == 0)
      throw StateError('Janus publisher handle is missing');
    return _publisherHandleId;
  }

  Future<List<(int, String?)>> joinAsPublisher({
    required int roomId,
    required String displayName,
  }) async {
    _requirePublisher();
    _roomId = roomId;
    final response = await _pluginRequest(
      handleId: _publisherHandleId,
      body: {
        'request': 'join',
        'room': roomId,
        'ptype': 'publisher',
        'display': displayName,
      },
    );
    final data = _mapAt(response, ['plugindata', 'data']);
    final publishers = data?['publishers'];
    if (publishers is! List) return const [];
    return publishers
        .whereType<Map>()
        .map((raw) {
          final map = raw.cast<Object?, Object?>();
          return ((map['id'] as num?)?.toInt() ?? 0, map['display'] as String?);
        })
        .where((value) => value.$1 != 0)
        .toList(growable: false);
  }

  Future<String> publishSdp(String offerSdp) async {
    _requirePublisher();
    final response = await _pluginRequest(
      handleId: _publisherHandleId,
      body: const {'request': 'configure', 'audio': true, 'video': true},
      jsep: {'type': 'offer', 'sdp': stripPrivateCandidates(offerSdp)},
    );
    final sdp = _mapAt(response, ['jsep'])?['sdp'] as String?;
    if (sdp == null || sdp.isEmpty)
      throw StateError('Janus SDP answer is missing');
    return sdp;
  }

  Future<String> subscribeToFeed(int feedId) async {
    if (_roomId == 0) throw StateError('Janus room is not joined');
    final handleId = await _attachSubscriberHandle();
    _subscriberHandles[feedId] = handleId;
    final response = await _pluginRequest(
      handleId: handleId,
      body: {
        'request': 'join',
        'room': _roomId,
        'ptype': 'subscriber',
        'feed': feedId,
      },
    );
    final sdp = _mapAt(response, ['jsep'])?['sdp'] as String?;
    if (sdp == null || sdp.isEmpty)
      throw StateError('Janus subscriber offer is missing');
    return sdp;
  }

  Future<void> answerSubscription({
    required int feedId,
    required String answerSdp,
  }) async {
    final handleId = _subscriberHandles[feedId];
    if (handleId == null) throw StateError('Subscriber handle is missing');
    await _pluginRequest(
      handleId: handleId,
      body: {'request': 'start', 'room': _roomId},
      jsep: {'type': 'answer', 'sdp': stripPrivateCandidates(answerSdp)},
    );
  }

  bool trickleIce({
    required int handleId,
    required String candidate,
    required String? sdpMid,
    required int sdpMLineIndex,
  }) {
    if (isPrivateCandidate(candidate)) return false;
    return _send({
      'janus': 'trickle',
      'session_id': _sessionId,
      'handle_id': handleId,
      'transaction': _transaction(),
      'candidate': {
        'candidate': candidate,
        if (sdpMid != null) 'sdpMid': sdpMid,
        'sdpMLineIndex': sdpMLineIndex,
      },
    });
  }

  bool trickleIceCompleted(int handleId) => _send({
    'janus': 'trickle',
    'session_id': _sessionId,
    'handle_id': handleId,
    'transaction': _transaction(),
    'candidate': const {'completed': true},
  });

  Future<void> leaveRoom() async {
    if (_sessionId == 0 || _publisherHandleId == 0) return;
    try {
      await _pluginRequest(
        handleId: _publisherHandleId,
        body: const {'request': 'leave'},
      );
    } catch (_) {
      // Disconnect still closes the authenticated Janus session.
    }
  }

  Future<int> _attachSubscriberHandle() async {
    final response = await _request({
      'janus': 'attach',
      'session_id': _sessionId,
      'plugin': 'janus.plugin.videoroom',
    });
    final id = _intAt(response, ['data', 'id']);
    if (id == 0) throw StateError('Janus subscriber handle is missing');
    return id;
  }

  Future<Map<String, Object?>> _pluginRequest({
    required int handleId,
    required Map<String, Object?> body,
    Map<String, Object?>? jsep,
  }) => _request({
    'janus': 'message',
    'session_id': _sessionId,
    'handle_id': handleId,
    'body': body,
    if (jsep != null) 'jsep': jsep,
  });

  Future<Map<String, Object?>> _request(Map<String, Object?> message) async {
    if (!_connected) throw StateError('Janus socket is not connected');
    final transaction = _transaction();
    final completer = Completer<Map<String, Object?>>();
    _pending[transaction] = completer;
    if (!_send({...message, 'transaction': transaction})) {
      _pending.remove(transaction);
      throw StateError('Janus request could not be sent');
    }
    try {
      return await completer.future.timeout(requestTimeout);
    } finally {
      _pending.remove(transaction);
    }
  }

  bool _send(Map<String, Object?> value) {
    final channel = _channel;
    if (channel == null) return false;
    try {
      channel.sink.add(jsonEncode(value));
      return true;
    } catch (_) {
      return false;
    }
  }

  void _handleMessage(dynamic raw) {
    if (raw is! String) return;
    try {
      final json = (jsonDecode(raw) as Map).cast<String, Object?>();
      final kind = json['janus'] as String?;
      final transaction = json['transaction'] as String?;
      if (transaction != null && kind != 'ack') {
        final completer = _pending[transaction];
        if (completer != null && !completer.isCompleted) {
          if (kind == 'error') {
            final reason =
                _mapAt(json, ['error'])?['reason'] ?? 'Unknown Janus error';
            completer.completeError(StateError(reason.toString()));
          } else {
            completer.complete(json);
          }
        }
      }
      if (kind == 'event') _handleEvent(json);
    } catch (_) {
      // Invalid unsolicited frames cannot mutate call state.
    }
  }

  void _handleEvent(Map<String, Object?> json) {
    final data = _mapAt(json, ['plugindata', 'data']);
    final publishers = data?['publishers'];
    if (publishers is List) {
      for (final raw in publishers.whereType<Map>()) {
        final feedId = (raw['id'] as num?)?.toInt();
        if (feedId != null) {
          _events.add(JanusPublisherJoined(feedId, raw['display'] as String?));
        }
      }
    }
    final left = (data?['unpublished'] ?? data?['leaving']) as num?;
    if (left != null) {
      final feedId = left.toInt();
      _subscriberHandles.remove(feedId);
      _events.add(JanusPublisherLeft(feedId));
    }
    final jsep = _mapAt(json, ['jsep']);
    final sdp = jsep?['sdp'] as String?;
    if (sdp == null) return;
    if (jsep?['type'] == 'answer') {
      _events.add(JanusRemoteAnswer(sdp));
    } else if (jsep?['type'] == 'offer') {
      final sender = (json['sender'] as num?)?.toInt();
      final feed = sender == null
          ? null
          : _subscriberHandles.entries
                .where((entry) => entry.value == sender)
                .firstOrNull
                ?.key;
      if (feed != null) _events.add(JanusRemoteOffer(feed, sdp));
    }
  }

  void _handleSocketError(Object error, [StackTrace? stack]) {
    _connected = false;
    for (final request in _pending.values) {
      if (!request.isCompleted) request.completeError(error, stack);
    }
    _pending.clear();
  }

  void _handleSocketDone() {
    _handleSocketError(StateError('Janus socket closed'));
  }

  void _requireSession() {
    if (_sessionId == 0) throw StateError('Janus session is not created');
  }

  void _requirePublisher() {
    _requireSession();
    if (_publisherHandleId == 0) throw StateError('VideoRoom is not attached');
  }

  String _transaction() => List.generate(
    12,
    (_) => 'abcdefghijklmnopqrstuvwxyz0123456789'[_random.nextInt(36)],
  ).join();

  Future<void> disconnect() async {
    _keepalive?.cancel();
    _keepalive = null;
    _connected = false;
    for (final request in _pending.values) {
      if (!request.isCompleted) {
        request.completeError(StateError('Janus client disconnected'));
      }
    }
    _pending.clear();
    _subscriberHandles.clear();
    await _subscription?.cancel();
    _subscription = null;
    await _channel?.sink.close(1000, 'Client disconnect');
    _channel = null;
    _sessionId = 0;
    _publisherHandleId = 0;
    _roomId = 0;
  }

  Future<void> dispose() {
    final active = _disposeTask;
    if (active != null) return active;
    _disposed = true;
    final operation = _dispose();
    _disposeTask = operation;
    return operation;
  }

  Future<void> _dispose() async {
    await disconnect();
    await _events.close();
  }

  static String stripPrivateCandidates(String sdp) => sdp
      .split(RegExp(r'\r?\n'))
      .where(
        (line) => !line.startsWith('a=candidate:') || !isPrivateCandidate(line),
      )
      .join('\r\n');

  static bool isPrivateCandidate(String candidate) => RegExp(
    r'\s(?:10\.\d+\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+|192\.168\.\d+\.\d+|169\.254\.\d+\.\d+|fe80:[0-9a-f:]+|[a-f0-9.-]+\.local)\s',
    caseSensitive: false,
  ).hasMatch(candidate);

  static Map<String, Object?>? _mapAt(
    Map<String, Object?> root,
    List<String> path,
  ) {
    Object? value = root;
    for (final part in path) {
      if (value is! Map) return null;
      value = value[part];
    }
    return value is Map ? value.cast<String, Object?>() : null;
  }

  static int _intAt(Map<String, Object?> root, List<String> path) {
    Object? value = root;
    for (final part in path) {
      if (value is! Map) return 0;
      value = value[part];
    }
    return (value as num?)?.toInt() ?? 0;
  }
}

extension<T> on Iterable<T> {
  T? get firstOrNull {
    final iterator = this.iterator;
    return iterator.moveNext() ? iterator.current : null;
  }
}
